sql/src/main/java/io/crate/planner/optimizer/rule/RewriteFilterOnOuterJoinToInnerJoin.java

```diff
@@ -0,0 +1,222 @@
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or
+ * implied.  See the License for the specific language governing
+ * permissions and limitations under the License.
+ *
+ * However, if you have executed another commercial license agreement
+ * with Crate these terms will supersede the license and you may use the
+ * software solely pursuant to the terms of the relevant commercial
+ * agreement.
+ */
+
+package io.crate.planner.optimizer.rule;
+
+import io.crate.analyze.WhereClause;
+import io.crate.analyze.relations.QuerySplitter;
+import io.crate.expression.eval.EvaluatingNormalizer;
+import io.crate.expression.operator.AndOperator;
+import io.crate.expression.symbol.FieldReplacer;
+import io.crate.expression.symbol.Literal;
+import io.crate.expression.symbol.Symbol;
+import io.crate.metadata.Functions;
+import io.crate.planner.node.dql.join.JoinType;
+import io.crate.planner.operators.Filter;
+import io.crate.planner.operators.LogicalPlan;
+import io.crate.planner.operators.NestedLoopJoin;
+import io.crate.planner.optimizer.Rule;
+import io.crate.planner.optimizer.matcher.Capture;
+import io.crate.planner.optimizer.matcher.Captures;
+import io.crate.planner.optimizer.matcher.Pattern;
+import io.crate.sql.tree.QualifiedName;
+
+import java.util.Map;
+import java.util.Set;
+
+import static io.crate.planner.optimizer.matcher.Pattern.typeOf;
+import static io.crate.planner.optimizer.matcher.Patterns.source;
+import static io.crate.planner.optimizer.rule.FilterOnJoinsUtil.getNewSource;
+import static io.crate.planner.optimizer.rule.FilterOnJoinsUtil.getRelationNames;
+
+/**
+ * If we can determine that a filter on an OUTER JOIN turns all NULL rows that the join could generate into a NO-MATCH
+ * we can push down the filter and turn the OUTER JOIN into an inner join.
+ *
+ * <p>
+ * So this tries to transform
+ * </p>
+ *
+ * <pre>
+ *     Filter (lhs.x = 1 AND rhs.x = 2)
+ *       |
+ *     NestedLoop (outerJoin)
+ *       /  \
+ *     LHS  RHS
+ * </pre>
+ *
+ * into
+ *
+ * <pre>
+ *     Filter
+ *       |
+ *     NestedLoop (innerJoin)
+ *       /      \
+ *   Filter      Filter
+ * (lhs.x = 1)    (rhs.x = 2)
+ *     |              |
+ *    LHS            RHS
+ * </pre>
+ *
+ * A case where this is *NOT* safe is for example:
+ *
+ * <pre>
+ * SELECT * FROM t1
+ *     LEFT JOIN t2 ON t1.t2_id = t2.id
+ * WHERE
+ *     coalesce(t2.x, 20) > 10   # becomes TRUE for null rows
+ * </pre>
+ *
+ *
+ * A case where the FILTER turns all null rows into no-matches:
+ *
+ * <pre>
+ * SELECT * FROM t1
+ *      LEFT JOIN t2 ON t1.t2_id = t2.id
+ * WHERE
+ *      t2.x = 10           # null = 10 -> null -> no match
+ * </pre>
+ */
+public final class RewriteFilterOnOuterJoinToInnerJoin implements Rule<Filter> {
+
+    private final Capture<NestedLoopJoin> nlCapture;
```

Comment:
is the restriction to nested loop artificial as equijoins could still be hash joins? maybe I'm missing something here but at least the rewritten inner join could possibly be a hash join

Reply:
HashJoins are already restricted to inner joins, so no rewrite on them would be necessary or possible.

But you're right that we could also do the equi-join detection on the result and possible transform them further to hash joins.

I'll follow up on this later as there are some things I'd like to work out:

a) `HashJoin` requires a `concreteRelation` argument (weird name, it seems to be the right side of the join) - the `NestedLoopJoin` doesn't have this attribute. It would be easy enough to add, but I want to first better understand why it is required in the first place. We currently do some "switch sides" logic when we build the execution plan. I think this optimization should be done on the operator level and I suspect it might be related to that.
b) hash joins can currently be turned off via session setting. I'd have to add the `SessionContext` or similar as argument to the rule application so that I could check it inside the rule. But I feel that going forward we should have session settings that either enable/disable a rule and that the check if a rule is enabled should be part of the optimizer - not inside the rule itself. That would advocate for having a separate `NL -> HashJoin` rule. Which would probably also fit better with the idea of integrating a cost calculator inside the rule framework. As a consequence of that we'd have to change the `JoinPlanBuilder` to initially build everything either as a generic `Join` or as `NestedLoop` and do the optimization as a second step via rules.

Reply:
Makes sense, thanks (I agree the optimizer should check if rules are enabled)